### ISSUE TITLE
Ref based transformers aren't safe for multithread use 

### ISSUE THREAD

**Issue by mgsloan** [fpco/monad-unlift#1]
As pointed out by https://github.com/fpco/mutable-containers, refs don't have any guarantees for multithreaded use.  It should probably be pointed out in the documentation for the "Control.Monad.Trans.*.Ref" modules that they also aren't thread safe.

I don't think there's a way around this, particularly due to `MonadState` not having a `modify` function as a class method.  However, `WriterRefT` could be made to not lose data passed to `tell` if it used the equivalent of `atomicModifyIORef`.

It's probably best to not change `WriterRefT` to be thread safe, because that would be inconsistent with the other `*RefT` types.

**Comment by snoyberg** [comment]
Doing so would preclude usage of the more efficient references like PRef, which doesn't have an atomic modification function. That use case may be rare enough that it's not too important.

**Comment by mgsloan** [comment]
Yup, I realized I left a "not" out of my last sentence, so it didn't make much sense.  My suggestion here is to add documentation about this, not necessarily to change the code.

**Comment by mitchellwrosen** [comment]
If you export `StateRefT(..)`, end users can write something like

``` haskell
getCAS :: StateRefT IORef s m (Ticket s)
getCAS = StateRefT (\ref -> liftBase (readForCAS ref))

putCAS :: Ticket s -> s -> StateRefT IORef s m (Bool, Ticket s)
putCAS ticket x = StateRefT (\ref -> liftBase (casIORef ref ticket x))
```

where the CAS ops are from https://hackage.haskell.org/package/atomic-primops-0.8.0.4/docs/Data-Atomics.html

Is this reason enough to export the newtype?

**Comment by mitchellwrosen** [comment]
Another reason to export the newtype: stepping a `StateRefT` computation forward manually without creating a new ref every time.
